examples/dnn_self_supervised_learning_ex.cpp

```diff
@@ -0,0 +1,366 @@
+            {
+                const auto idx = rnd.get_random_32bit_number() % training_images.size();
+                auto image = training_images[idx];
+                batch.emplace_back(augment(image, false, rnd), augment(image, true, rnd));
+            }
+            trainer.train_one_step(batch);
+
+            // Compute the empirical cross-correlation matrix every 100 steps. Again,
+            // this is not needed for the training to work, but it's nice to visualize.
+            if (trainer.get_train_one_step_calls() % 100 == 0)
+            {
+                // Wait for threaded processing to stop in the trainer.
+                trainer.get_net(force_flush_to_disk::no);
+                // Get the output from the last fc layer
+                const auto& out = net.subnet().get_output();
+                // The trainer might have synchronized its state to the disk and cleaned
+                // the network state. If that happens, the output will be empty, in
+                // which case, we just skip the empirical cross-correlation matrix
+                // computation.
+                if (out.size() == 0)
+                    continue;
+                // Separate both augmented versions of the images
+                alias_tensor split(out.num_samples() / 2, dims);
+                auto za = split(out);
+                auto zb = split(out, split.size());
+                gamma = 1;
+                beta = 0;
+                // Perform batch normalization on each feature representation, independently.
+                tt::batch_normalize(eps, za_norm, means, invstds, 1, rms, rvs, za, gamma, beta);
+                tt::batch_normalize(eps, zb_norm, means, invstds, 1, rms, rvs, za, gamma, beta);
+                // Compute the empirical cross-correlation matrix between the features and
+                // visualize it.
+                tt::gemm(0, eccm, 1, za_norm, true, zb_norm, false);
+                eccm /= batch_size;
+                matrix<unsigned char> c_img;
+                assign_image(c_img, abs(mat(eccm)) * 255);
+                win.set_image(c_img);
+                win.set_title("Barlow Twins step#: " + to_string(trainer.get_train_one_step_calls()));
+            }
+        }
+        trainer.get_net();
+        net.clean();
+        // After training, we can discard the projector head and just keep the backone
+        // to train it or finetune it on other downstream tasks.
+        serialize("resnet50_self_supervised_cifar_10.net") << layer<5>(net);
+    }
+
+    // To check the quality of the learned feature representations, we will train a linear
+    // classififer on top of the frozen backbone.
+    model::infer inet;
+    // Assign the network, without the projector, which is only used for the self-supervised
+    // training.
+    layer<2>(inet) = layer<5>(net);
+    // Freeze the backbone
+    set_all_learning_rate_multipliers(layer<2>(inet), 0);
```

Comment:
You use the same augmentation.  You just run the backbone on however many augmented images you want to use as your training set.  Then run the SVM solver on that training set. Like say this training loop here generates 10000 augmented images during training, you could make 10000 augmented images ahead of time and then run the solver on them.  But really you probably need less than there are SGD iterations, since SGD is an awful solver.  The sample efficiency of SGD is really bad.  Anyway, you should try it and see what I mean.

Reply:
I've been re-running the example again with the svm multiclass linear trainer (I deleted the sync files, so I have to run it from scratch again...)
I will try with and without data augmentation and probably make another PR.
Since the CIFAR-10 dataset is quite small, it's feasible to keep all the generated features in memory, but for datasets with millions of images... I don't know... From what I've seen the SVM multiclass linear trainer needs all samples in memory right?

Reply:
Right, they all have to be in memory. 

There is only so much capacity in a little linear model like this. At some point it's got all the data and more doesn't change the result. So for really large datasets you likely don't need to load it all anyway.